staging/src/k8s.io/apiserver/pkg/server/healthz.go

```diff
@@ -18,28 +18,117 @@ package server
 
 import (
 	"fmt"
-
 	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"net/http"
+	"time"
 )
 
-// AddHealthzCheck allows you to add a HealthzCheck.
+// AddHealthzCheck adds HealthzCheck(s) to both healthz and readyz. All healthz checks
+// are automatically added to readyz, since we want to avoid the situation where the
+// apiserver is ready but not live.
 func (s *GenericAPIServer) AddHealthzChecks(checks ...healthz.HealthzChecker) error {
 	s.healthzLock.Lock()
 	defer s.healthzLock.Unlock()
 
-	if s.healthzCreated {
+	if s.livenessHealthChecksLocked || s.readinessHealthChecksLocked {
 		return fmt.Errorf("unable to add because the healthz endpoint has already been created")
 	}
+	s.livenessHealthChecks = append(s.livenessHealthChecks, checks...)
+	s.readinessHealthChecks = append(s.readinessHealthChecks, checks...)
+	return nil
+}
+
+// AddReadyzChecks allows you to add a HealthzCheck to readyz.
+func (s *GenericAPIServer) AddReadyzChecks(checks ...healthz.HealthzChecker) error {
```

Comment:
if we use "health check" as the general term, better drop the `z` everywhere where don't mean the actual `/healthz` endpoint